Meta: disable features with a click in the options

bfred-it opened the pull request:
Closes https://github.com/sindresorhus/refined-github/issues/2040

<img width="641" alt="Screenshot 2019-05-19 at 18 30 14" src="https://user-images.githubusercontent.com/1402241/57980980-ab411c00-7a64-11e9-960e-e35865bd4254.png">

<details>
<summary>Outdated notes</summary>


Notes:

- I had initially had the link on the feature name, but a bunch of blue links looked tacky.
- I thought we could handle this with an autocomplete-like feature, but with that we can't _also_ link to the source (this would probably be solved if we had descriptions in each file)
- ~~about _"checked means disabled":_ this is because the features are enabled by default, but to show them as _"checked means enabled"_ we'd have to specify them all in the `defaults` in `background.js`.~~ <br><br> ~~The form could be made more understandable by using colors, i.e. green on unchecked and red on checked boxes.~~ Edit: I think the strikethrough is clear

<table><tr><td><img width="457" alt="options with list" src="https://user-images.githubusercontent.com/1402241/52787594-52469e00-3099-11e9-98aa-c94e2687e8fc.png"></td></tr></table>

<details><summary>screenshot of bdea47d (outdated)</summary>
<table><tr><td><img width="457" alt="options with list" src="https://user-images.githubusercontent.com/1402241/52786648-5fae5900-3096-11e9-9597-3c7e96fb81b0.png"></td></tr></table></details>


</details>

sindresorhus commented:
> * but to show them as _"checked means enabled"_ we'd have to specify them all in the `defaults` in `background.js`.

Can we not send the list of all rule IDs to `background.js` and set the defaults?

bfred-it commented:
> Can we not send the list of all rule IDs to `background.js` and set the defaults?

Good point. Then we can store them as regular checkboxes instead of combining them as a single property (https://github.com/bfred-it/webext-options-sync/pull/16)

bfred-it commented:
<img width="411" alt="Screenshot 2019-05-19 at 18 28 14" src="https://user-images.githubusercontent.com/1402241/57980957-85b41280-7a64-11e9-9d41-7b1572f48798.png">

Also looks good in Firefox:

<img width="641" alt="Screenshot 2019-05-19 at 18 30 14" src="https://user-images.githubusercontent.com/1402241/57980956-85b41280-7a64-11e9-92a8-f164572d93ec.png">

bfred-it commented:
Moved description to second line, added screenshots field.

<img width="532" alt="" src="https://user-images.githubusercontent.com/1402241/57981424-e6921980-7a69-11e9-9ff4-b176f1f5ff34.png">

bfred-it commented on source/background.ts:
```diff
@@ -3,27 +3,41 @@ import {addContextMenu} from 'webext-domain-permission-toggle';
+	customCSS: string;
+	personalToken: string;
+	logging: boolean;
+	[featureName: string]: string | boolean;
```
This index signature is required because we're adding a lot of `'feature:XXX': boolean` to the options, but it's causing some trouble. More info in the next comments.

cc @nickytonline

nickytonline replied:
I'll take a look later tonight.

bfred-it commented on source/libs/features.tsx:
```diff
@@ -98,24 +91,19 @@ const globalReady: Promise<GlobalOptions> = new Promise(async resolve => {
+	log = options.logging ? console.log : () => { };
 
-	resolve(options);
+	resolve(options as Options);
```
```
[tsl] ERROR in source/libs/features.tsx(103,10)
      TS2345: Argument of type 'Record<string, any>' is not assignable to parameter of type 'Options | PromiseLike<Options> | undefined'.
  Property 'then' is missing in type 'Record<string, any>' but required in type 'PromiseLike<Options>'.
```

nickytonline replied:
```suggestion
	resolve(options);
```

bfred-it commented on source/libs/features.tsx:
```diff
@@ -98,24 +91,19 @@ const globalReady: Promise<GlobalOptions> = new Promise(async resolve => {
-	if (options.customCSS.trim().length > 0) {
+	const options = await new OptionsSync().getAll();
+
+	if ((options.customCSS as string).trim().length > 0) {
```
```
[tsl] ERROR in source/libs/features.tsx(96,24)
      TS2339: Property 'trim' does not exist on type 'string | number | boolean'.
  Property 'trim' does not exist on type 'number'.
```

bfred-it commented on source/manifest.json:
```diff
@@ -34,6 +34,7 @@
 	"background": {
 		"scripts": [
 			"browser-polyfill.min.js",
+			"content.js",
```
This is causing the background script to fail because some features have code outside `init`. I'll have to fix it

nickytonline commented on source/libs/features.tsx:
```diff
@@ -98,24 +91,19 @@ const globalReady: Promise<GlobalOptions> = new Promise(async resolve => {
-	if (options.customCSS.trim().length > 0) {
+	const options = await new OptionsSync().getAll();
+
+	if ((options.customCSS as string).trim().length > 0) {
```
```suggestion
	if (options.customCSS.trim().length > 0) {
```

bfred-it replied:
`as string` has been introduced to avoid that error. If I drop `as string` I get 

```
TS2339: Property 'trim' does not exist on type 'string | number | boolean'.
  Property 'trim' does not exist on type 'number'.
```

nickytonline replied:
It's fine if you cast as I mentioned above, `const options = (await new OptionsSync().getAll()) as Options;`.

nickytonline commented on source/libs/features.tsx:
```diff
@@ -98,24 +91,19 @@ const globalReady: Promise<GlobalOptions> = new Promise(async resolve => {
-	};
-
-	if (options.customCSS.trim().length > 0) {
+	const options = await new OptionsSync().getAll();
```
```suggestion
	const options = (await new OptionsSync().getAll()) as Options;
```

This `as` can probably be avoided if we improve the types for `OptionsSync`.

nickytonline replied:
I wasn't sure how to suggest changes for multiple lines, hence the 3 one liner suggested changes. 🙃

bfred-it replied:
This just moves the `as Options` but I'd rather avoid it

nickytonline replied:
Moving it prevents a cast to `string` when trimming. What would be interesting is to improve the types of `OptionsSync`. Making `getAll` generic, would avoid any casting.

bfred-it replied:
I see what you mean now.

> Making getAll generic, would avoid any casting

👍

bfred-it commented on source/background.ts:
```diff
@@ -68,3 +82,12 @@ browser.runtime.onInstalled.addListener(async ({reason}) => {
+
+// @ts-ignore because this is only needed sometimes
+function featureWasRenamed(from: string, to: string): Migration { // eslint-disable-line @typescript-eslint/no-unused-vars
+	return (options: AnyObject) => {
```
Forgot this

It should be:

```suggestion
	return (options: Options) => {
```

but:

```
TS2322: Type '(options: Options) => void' is not assignable to type 'Migration'.
  Types of parameters 'options' and 'savedOptions' are incompatible.
    Type 'Record<string, string | number | boolean>' is missing the following properties from type 'Options': customCSS, personalToken, logging
```

nickytonline replied:
Just heading to bed, but I looked at the webex-options-sync types. We can probably update the types for `Migration` to be generic. Maybe [something like this](https://typescript-play.js.org/#code/PQKhCgAIUhBSC2BDADpA9gMwygLgS3QDsBnSJMk3AJ3yIHMz1rIAjddAGwFMlSA6SABUAFt0gBrbgE8yAd3ydOkEUgBu43OkRJcAYxGRcYyJmYJT+bpwAmJAOSQABkSQJuT8rhr5WAV1xuEn4oEGBwbgAPFGZcI2kUcQB5PEJSSABeSAAlbj1mGwAeKloGABpIErp6SAAfSCI-BFZuFnr2Ll4iAD4AbnBwUAhoSAAJPhseFhJ8eldcP2pxMxY5MSJyDajA0jTIPxQbXSCQ6HComOo4ukDqTCQ9cQBZWepdNMKhFIJiMm3uIh2SAAbwAvplIN80iRuiDwABIAAUJHU3BsUN+AC5hBjSBUbNx7n5OLgSNivqlfgBKbFqdD4Gz9UEDG6te6PHIAcVGuLIwIRej8VHQCAAwgBlcXYqoMfrwxLUEjEJCcIToKREaU+WUIzjoej0arYjo8PhMgb5UhxQJUABifiIeh+msgL3ob2dhWy3N5sKyiPQlLxkAJRJJJCpmVh-MgscglqVPH4evoiPs0nQ9ip5oGQA). The only thing I don't like about this is the options can literally be anything then as long as it's an object.

bfred-it replied:
I think actually the types of `OptionsSync.Options` should be `typeof defaults`, which is the definition object passed by the user. I don't know if this can be passed around through the instance.

In the [tasks here](https://github.com/bfred-it/webext-options-sync/issues/5) there are more info about what I'm planning to do on that extension as well, I'll mention you when I have that done first.

bfred-it commented:
This should be good to go

notlmn commented:
For some unusual reason Chrome sets a `line-height` of `1.8em` on paragraphs, can we set it to `1.5em`.

Just a nit-pick. Please 🙏.

notlmn commented on source/libs/features.tsx:
```diff
@@ -139,6 +126,13 @@ const getShortcuts = (): Shortcut[] => [...shortcutMap.values()];
 const add = async (definition: FeatureDetails): Promise<void> => {
+	// In chrome:// pages, just collect the features in a global variable
+	if (!location.protocol.startsWith('http')) {
+		window.collectFeatures = window.collectFeatures || new Map();
```
Why not instantiate `collectFeatures` on `window` outside `add`?

https://github.com/sindresorhus/refined-github/pull/1780/files#diff-9f662d0cb609bc9bff7176ed8ee00bb6R29 is always throwing error on page load for me.

![image](https://user-images.githubusercontent.com/37769974/58177945-eb053f00-7cc2-11e9-8098-90c58623ba68.png)

busches replied:
Same, just came to mention this.

busches commented:
Everything works on GHE that I can tell, only change I would have is a way to quickly collapse the Feature list in options to easily get to the Custom CSS option.

simonkotwicz commented:
You could also just move the custom CSS box above the feature list. If you wanted the feature list to be more compact you could also add the descriptions as tooltips instead. Just my two cents.

bfred-it commented:
> For some unusual reason Chrome sets a `line-height` of `1.8em` on paragraphs, can we set it to `1.5em`.

I don't see a problem with that, we try to leave the style to the browsers as much as possible.

bfred-it commented:
> a way to quickly collapse the Feature list in options to easily get to the Custom CSS option.

We could move it to a tab, but I'd like to keep the options as "flat" as possible, also so we don't have to care much for their styles (given that it's shared between Chrome and Firefox)

> You could also just move the custom CSS box above the feature list

Yeah, I like this better. I left the "logging" toggle below the list because it's more of a development feature anyway.

> you could also add the descriptions as tooltips instead

Not tooltips because descriptions are long and sometimes include `code`, but I wouldn't mind moving the description behind an `expand` button.

I'd hide the feature name too, but it gives users a handle to remember, discuss and find.

simonkotwicz commented:
This is out of the scope for this PR but here are some other thoughts I had:
- Group features in the same way they are grouped in the README (Declutter, More actions, etc.). However, this would require extra information to be added to each feature and maybe some feature don't even fit in those categories.
- Mark new features with a different colour or "NEW" tag (or brought to the top of the list under a "New" heading) until the user clicks on the feature/source/screenshot.

bfred-it commented:
I need to work on this further because the migrations aren't being triggered at times, I suspect it's because `setTimeout` in `options-storage` makes `webext-options-sync` call `chrome.runtime.onInstalled` too late.

I think the solution is to extract the features list at build time. Thankfully I had partially wrote this before.

notlmn commented on source/options.tsx:
```diff
@@ -1,8 +1,56 @@
+}
+
+const featureCheckboxes = [...window.collectFeatures.entries()]
+	.sort(([a], [b]) => a.localeCompare(b)) // Sort bt feature name
```
```suggestion
	.sort(([a], [b]) => a.localeCompare(b)) // Sort by feature name
```
